rfcs/43-many-worlds.md

```diff
@@ -0,0 +1,522 @@
+
+   let mut best_action = Action::Pass;
+   let mut best_score = Score(0);
+
+   for evaluation in eval_events.iter(){
+      if evaluation.score >= best_score {
+         best_action = evaluation.action;
+      }
+   }
+
+   action_events.send(best_action);
+}
+```
+
+### Advanced: custom strategies for multiple worlds
+
+You can modify the default strategy described above by setting up your own custom runner, which can be used to control the execution of schedules in arbitrary ways.
+
+Note that schedules are initialized on particular worlds, and will not run on a world with a mismatched `WorldLabel`.
+
+## Implementation strategy
+
+`App`'s new struct under this proposal would be:
+
+```rust
+pub struct App {
+ // WorldId is not very readable; we can steal labelling tech from systems
+ pub worlds: HashMap<WorldLabel, World>,
+  // Each Schedule must have a `World`, but not all worlds need schedules
+ pub schedules: HashMap<WorldLabel, Schedule>,
+ // Controls where app
+ // This is important to preserve the current ergonomics in the simple case
+ // Defaults to `CoreWorld::Main`
+ pub current_world: Box<dyn WorldLabel>,
+  // This is needed in order to correctly hand out read-only access
+ pub global_world: Box<dyn WorldLabel>,
+ // Which world the entities to be used for rendering and sound come from
+ // The relevant components will be automatically copied and moved 
+ // to the appropriate worlds in a pipelined fashion
+ // Defaults to `CoreWorld::Main`
+ pub source_world: Box<dyn WorldLabel>,
+ // Exactly the same as before
+ pub runner: Box<dyn Fn(App) + 'static, Global>,
+}
+```
+
+The standard main loop would have the following basic structure:
+
+1. Create a `Global<ComputeTaskPool>`, which all worlds can access when they need threads to run their systems.
+2. For each `WorldLabel` in `app.world.keys()`, initialize the corresponding schedule with the correct label.
+3. For each world, run the corresponding schedule if it exists.
+   1. Each world is fully independent, and each scheduler can request more threads as needed in a greedy fashion.
+4. Run the global schedule.
+5. At the end of each loop, put the entire `App` back together and run any `AppCommands`.
+   1. Be sure to apply standard `Commands` to each world.
+6. Go to 2.
+
+This should be programmed using the default (and minimal) runners.
+Custom schedule and synchronization behavior can be added with a custom runner.
+
+### Cloneable schedules
+
+Schedules must be cloneable in order to reasonably spin up new worlds.
+
+When schedules are cloned, they must become uninitialized again.
+
+### AppCommands
+
+`AppCommands` directly parallel `Commands` API and design where possible, likely abstracting out a trait.
+
+They operate on `&mut App`.
+
+The initial set of `AppCommands` needed for this RFC are:
+
+- `spawn_world(label: impl WorldLabel)`
+- `despawn_world(label: impl WorldLabel)`
+- `set_world(label: impl WorldLabel)`
```

Comment:
i am wary of builder APIs which change the meaning of downstream methods in this way. usually i would go with something like

```rust
app.with_world(label, |&mut app| app.add_plugin(MyPlugin))
```

but i maybe there are backwards compat concerns?